File: VB/GridExample/EmptyGridHelper.vb
Imports System.Windows
Imports DevExpress.Xpf.Grid
Imports DevExpress.Xpf.Core.Native
Imports System.Windows.Controls
Imports System.Windows.Data

Namespace GridExample

    Public Class EmptyGridHelper

        Public Shared ReadOnly HasVisibleRowsProperty As DependencyProperty = DependencyProperty.RegisterAttached("HasVisibleRows", GetType(Boolean), GetType(EmptyGridHelper), New PropertyMetadata(False))

        Public Shared ReadOnly EmptyGridTemplateProperty As DependencyProperty = DependencyProperty.RegisterAttached("EmptyGridTemplate", GetType(DataTemplate), GetType(EmptyGridHelper), New UIPropertyMetadata(Nothing, New PropertyChangedCallback(AddressOf EmptyGridTemplateChanged)))

        Private Shared Sub EmptyGridTemplateChanged(ByVal sender As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            Dim view As TableView = TryCast(sender, TableView)
            Dim emptyGridTemplate As DataTemplate = CType(e.NewValue, DataTemplate)
            If view Is Nothing OrElse emptyGridTemplate Is Nothing Then Return
            AddHandler view.Grid.PropertyChanged, New System.ComponentModel.PropertyChangedEventHandler(AddressOf Grid_PropertyChanged)
            AddHandler view.Grid.ItemsSourceChanged, New ItemsSourceChangedEventHandler(AddressOf Grid_ItemsSourceChanged)
            AddHandler view.Loaded, New RoutedEventHandler(AddressOf view_Loaded)
        End Sub

        Private Shared Sub view_Loaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
            Dim view As TableView = TryCast(sender, TableView)
            Dim emptyGridTemplate As DataTemplate = CType(view.GetValue(EmptyGridTemplateProperty), DataTemplate)
            If view Is Nothing OrElse emptyGridTemplate Is Nothing Then Return
            AddEptyGridContentPresenter(view, emptyGridTemplate)
        End Sub

        Private Shared Sub AddEptyGridContentPresenter(ByVal view As TableView, ByVal emptyGridTemplate As DataTemplate)
            Dim emptyGridContentresenter As ContentPresenter = New ContentPresenter()
            emptyGridContentresenter.ContentTemplate = emptyGridTemplate
            emptyGridContentresenter.HorizontalAlignment = HorizontalAlignment.Stretch
            emptyGridContentresenter.VerticalAlignment = VerticalAlignment.Stretch
            Dim emptyGridContentVisibilityBinding As Binding = New Binding("(local:EmptyGridHelper.HasVisibleRows)") With {.Source = view}
            emptyGridContentresenter.SetBinding(ContentPresenter.VisibilityProperty, emptyGridContentVisibilityBinding)
            Dim container As Grid = CType(LayoutHelper.FindElementByName(view, "rowPresenterGrid"), Grid)
            container.Children.Add(emptyGridContentresenter)
        End Sub

        Private Shared Sub Grid_ItemsSourceChanged(ByVal sender As Object, ByVal e As ItemsSourceChangedEventArgs)
            If e.NewDataSource Is Nothing Then
                CType(sender, GridControl).View.SetValue(HasVisibleRowsProperty, False)
            End If
        End Sub

        Private Shared Sub Grid_PropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs)
            If Equals(e.PropertyName, "VisibleRowCount") Then
                Dim grid As GridControl = CType(sender, GridControl)
                grid.View.SetValue(HasVisibleRowsProperty, grid.VisibleRowCount <> 0)
            End If
        End Sub

'#Region "CLRs"
        Public Shared Function GetHasVisibleRows(ByVal obj As DependencyObject) As Boolean
            Return CBool(obj.GetValue(HasVisibleRowsProperty))
        End Function

        Public Shared Sub SetHasVisibleRows(ByVal obj As DependencyObject, ByVal value As Boolean)
            obj.SetValue(HasVisibleRowsProperty, value)
        End Sub

        Public Shared Function GetEmptyGridTemplate(ByVal obj As DependencyObject) As DataTemplate
            Return CType(obj.GetValue(EmptyGridTemplateProperty), DataTemplate)
        End Function

        Public Shared Sub SetEmptyGridTemplate(ByVal obj As DependencyObject, ByVal value As DataTemplate)
            obj.SetValue(EmptyGridTemplateProperty, value)
        End Sub
'#End Region
    End Class
End Namespace

File: CS/GridExample/EmptyGridHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Windows;
using DevExpress.Xpf.Grid;
using DevExpress.Xpf.Core.Native;
using System.Windows.Controls;
using System.Windows.Shapes;
using System.Windows.Data;
using DevExpress.Xpf.Core;

namespace GridExample {
    public class EmptyGridHelper {

        public static readonly DependencyProperty HasVisibleRowsProperty =
            DependencyProperty.RegisterAttached("HasVisibleRows", typeof(bool), typeof(EmptyGridHelper), new PropertyMetadata(false));

        public static readonly DependencyProperty EmptyGridTemplateProperty =
            DependencyProperty.RegisterAttached("EmptyGridTemplate", typeof(DataTemplate), typeof(EmptyGridHelper), new UIPropertyMetadata(null, new PropertyChangedCallback(EmptyGridTemplateChanged)));

        static void EmptyGridTemplateChanged(DependencyObject sender, DependencyPropertyChangedEventArgs e) {
            TableView view = sender as TableView;
            DataTemplate emptyGridTemplate = (DataTemplate)e.NewValue;
            if (view == null || emptyGridTemplate == null) return;

            view.Grid.PropertyChanged += new System.ComponentModel.PropertyChangedEventHandler(Grid_PropertyChanged);
            view.Grid.ItemsSourceChanged += new ItemsSourceChangedEventHandler(Grid_ItemsSourceChanged);

            view.Loaded += new RoutedEventHandler(view_Loaded);
        }

        static void view_Loaded(object sender, RoutedEventArgs e) {
            TableView view = sender as TableView;

            DataTemplate emptyGridTemplate = (DataTemplate)view.GetValue(EmptyGridHelper.EmptyGridTemplateProperty);
            if (view == null || emptyGridTemplate == null) return;

            AddEptyGridContentPresenter(view, emptyGridTemplate);
        }

        static void AddEptyGridContentPresenter(TableView view, DataTemplate emptyGridTemplate) {
            ContentPresenter emptyGridContentresenter = new ContentPresenter();
            emptyGridContentresenter.ContentTemplate = emptyGridTemplate;

            emptyGridContentresenter.HorizontalAlignment = HorizontalAlignment.Stretch;
            emptyGridContentresenter.VerticalAlignment = VerticalAlignment.Stretch;

            Binding emptyGridContentVisibilityBinding = new Binding("(local:EmptyGridHelper.HasVisibleRows)") {
                Source = view,
                Converter = new BoolToVisibilityInverseConverter() { }
            };
            emptyGridContentresenter.SetBinding(ContentPresenter.VisibilityProperty, emptyGridContentVisibilityBinding);

            Grid container = (Grid)LayoutHelper.FindElementByName(view, "rowPresenterGrid");

            container.Children.Add(emptyGridContentresenter);
        }


        static void Grid_ItemsSourceChanged(object sender, ItemsSourceChangedEventArgs e) {
            if (e.NewDataSource == null) {
                ((GridControl)sender).View.SetValue(EmptyGridHelper.HasVisibleRowsProperty, false);
            }
        }

        static void Grid_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e) {
            if (e.PropertyName == "VisibleRowCount") {
                GridControl grid = (GridControl)sender;
                grid.View.SetValue(EmptyGridHelper.HasVisibleRowsProperty, grid.VisibleRowCount != 0);
            }
        }

        #region CLRs
        public static bool GetHasVisibleRows(DependencyObject obj) {
            return (bool)obj.GetValue(HasVisibleRowsProperty);
        }

        public static void SetHasVisibleRows(DependencyObject obj, bool value) {
            obj.SetValue(HasVisibleRowsProperty, value);
        }

        public static DataTemplate GetEmptyGridTemplate(DependencyObject obj) {
            return (DataTemplate)obj.GetValue(EmptyGridTemplateProperty);
        }

        public static void SetEmptyGridTemplate(DependencyObject obj, DataTemplate value) {
            obj.SetValue(EmptyGridTemplateProperty, value);
        }
        #endregion  

    }
}
